### DISCUSSION TITLE
wasm 'threading' demo / sample

### DISCUSSION THREAD

**Discussion by pm100** [emilk/egui#4798] [General]
I am working on an app that
- will do 'chunky ' bits of processing 
- should run on desktop and wasm

There does not seem to be a 'blessed' solution to this. It would be awesome if somebody could contribute an example that could be put in the demo app. Dont know if it would threads or async or web worker or whatever combination thereof.
I can be the only one with this issue.

I would be happy to take raw code and abstract it into a demo

**Comment by emilk** [comment]
You can look at how `ehttp` does this. Here is the first bread crumb: https://github.com/emilk/ehttp/blob/4e63d731298d5f1411eb96a1f4e480d8bd771d50/example_eframe/src/app.rs#L65-L68

**Reply by pm100** [reply]
That is a wrapper around asking javascript to do http, ie real async stuff, neat but I dont think its what I need. I want to run chunky rust code, I have a compiler / assembler I need to run. Its a fast library but calling it inline will stall the UI for a few seconds. Any other pointers?

**Reply by emilk** [reply]
You are not allowed to stall the UI thread in the browser, so you need to use WebWorkers.

**Reply by pm100** [reply]
yup - I know. I just wondered if anybody had any examples here. Given that egui support web and desktop - I expect it to be a fairly common issue

**Reply by ghost** [reply]
Hi, what I did is something like:

```
#[cfg(target_arch = "wasm32")]
impl ApiCommandSender for WasmApiSender {
    fn send_command(&mut self, api_command: ApiCommand) -> () {
        wasm_bindgen_futures::spawn_local(async move {
            process_api_command(&api_command).await;
        });
    }
}

```

In the `process_api_command` I am using reqwest to make the actual get to the server.  The response is send back over a channel that is encapsulated in the api_command struct:

```
            let response: Result<reqwest::Response, reqwest::Error> = client.get(url).send().await;
            // stuff and send back the result over a channel
            response_sender.send(result).await.unwrap();
```

If you only need wasm than the channel part is not required, you could just await. I needed it in my app for cross wasm / native api sending.

HTH
